lib/internal/test_runner/test.js

```diff
@@ -76,15 +78,42 @@ const { testNamePatterns, testOnlyFlag } = parseCommandLine();
 let kResistStopPropagation;
 
 function stopTest(timeout, signal) {
+  const deferred = createDeferredPromise();
+  const abortListener = addAbortListener(signal, deferred.resolve);
+  let timer;
+  let disposeFunction;
+
   if (timeout === kDefaultTimeout) {
-    return once(signal, 'abort');
+    disposeFunction = abortListener[SymbolDispose];
+  } if (timeout !== kDefaultTimeout) {
+    timer = setTimeout(() => deferred.resolve(), timeout);
+    timer.unref();
+
+    ObjectDefineProperty(deferred, 'promise', {
+      __proto__: null,
+      configurable: true,
+      writable: true,
+      value: PromisePrototypeThen(deferred.promise, () => {
+        throw new ERR_TEST_FAILURE(
+          `test timed out after ${timeout}ms`,
+          kTestTimeoutFailure,
+        );
+      }),
```

Comment:
I think this would not act as expected. If `signal` aborts, we still enter this `.then`, which is probably not what we want (as this used to only happen on timeout)

Perhaps you could add a UT ensuring when signal aborted we get expected error even when a timeout was provided? 🙂

Reply:
This could probably be solved by having 2 separate promises and return a `PromiseRace` etc

Reply:
@rluvaton Did you check this behaves as expected?

Reply:
yes, checked and it behave as expected, it still throws abort error and not this

Reply:
Just making sure we are talking about the same thing.
In a test which has a `timeout` option, and the `signal` is aborted prior to `timeout` - we are getting the expected error and not the `test timed out after ${timeout}ms`?

Reply:
Yes